text/0000-autoreborrow-traits.md

```diff
@@ -0,0 +1,468 @@
+- Usage: Provide exclusive access to statically knowable (or mostly knowable)
+  hardware peripherals in an embedded system codebase.
+
+### [GcScope and NoGcScope](https://github.com/trynova/nova/blob/260683fccbad913491c758781965eb3bf67b531e/nova_vm/src/engine/context.rs#L40-L78)
+
+- Size: ZST.
+- Kind: Marker types with no runtime existence. `GcScope` has two lifetimes, one
+  lifetime is exclusive and one is shared. `NoGcScope` has two shared lifetimes.
+  The always-shared lifetime is not very relevant to reborrowing and can be
+  ignored here.
+- Usage: `GcScope` provides exclusive access to a garbage collector, while its
+  shared reborrow `NoGcScope` is used to create handles to garbage collected
+  data. Handles are thus invalidated if `GcScope` is reborrowed as exclusive (or
+  moved).
+
+### [MutableHandle](https://doc.servo.org/mozjs/gc/struct.MutableHandle.html) and [Handle](https://doc.servo.org/mozjs/gc/struct.Handle.html)
+
+- Size: Pointer-sized.
+- Kind: `MutableHandle` is conceptually an exclusive reference type. `Handle` is
+  a wrapper around a shared reference.
+- Usage: `MutableHandle` is a reference type wrapping a raw pointer that can
+  unsafely be turned into an exclusive reference on the condition that the
+  garbage collector does not run while the exclusive reference exists. `Handle`
+  is a plain shared reference wrapper; the pointed-to value is either a pointer
+  to garbage collected data, is internally mutable, or is never mutated by the
+  garbage collector. (Or so I assume.)
+
+### [Mat](https://docs.rs/faer/latest/faer/mat/struct.Mat.html), [MatMut](https://docs.rs/faer/latest/faer/mat/struct.MatMut.html), [MatRef](https://docs.rs/faer/latest/faer/mat/struct.MatRef.html), [ColMut](https://docs.rs/faer/latest/faer/col/struct.ColMut.html), and [ColRef](https://docs.rs/faer/latest/faer/col/struct.ColRef.html)
+
+`Mat` is an owned rectangular matrix while `MatMut`, `MatRef`, `ColMut`, and
+`ColRef` are various kinds of custom reference types over the source `Mat`.
+
+- Size: Usually 5 pointer-sizes for `MatMut` and `MatRef`, 3 for `ColMut` and
+  `ColRef`.
+- Kind: `MatMut` and `ColMut` are exclusive reference types. `MatRef` and
+  `ColRef` are shared reference types.
+- Usage: Enable various near-zero-cost operations over matrices by binding the
+  iteration to the original `Mat` through the reference types while statically
+  ensuring that indexed access is not required.
+
+Note: `Mat` is also reborrowable into `MatMut` and `MatRef`, but also has a
+`Drop` implementation. If reborrowing is understood to act on lifetimes of
+reference-like types, then reborrowing `Mat` is outside that definition as it
+neither has a lifetime nor is a reference-like type. It could be argued that
+`Mat` does have an implied `'static` lifetime, but the lack reference-like
+aspects of it at least cannot be denied. It might be that this use-case would be
+better understood as a generalised `AsRef` implementation.
```

Comment:
I think more `Deref[Mut]` than `AsRef`, since it's implicit coercion? `Mat` feels out of scope except to the extent that some generalized coercion trait might capture both reborrowing and deref.